webrisk/webrisk_client.go

```diff
@@ -0,0 +1,522 @@
+	// service. This field is required.
+	APIKey string
+
+	// ID and Version are client metadata associated with each API request to
+	// identify the specific implementation of the client.
+	// They are similar in usage to the "User-Agent" in an HTTP request.
+	// If empty, these default to DefaultID and DefaultVersion, respectively.
+	ID      string
+	Version string
+
+	// DBPath is a path to a persistent database file.
+	// If empty, WebriskClient operates in a non-persistent manner.
+	// This means that blacklist results will not be cached beyond the lifetime
+	// of the WebriskClient object.
+	DBPath string
+
+	// UpdatePeriod determines how often we update the internal list database.
+	// If zero value, it defaults to DefaultUpdatePeriod.
+	UpdatePeriod time.Duration
+
+	// ThreatLists determines which threat lists that WebriskClient should
+	// subscribe to. The threats reported by LookupURLs will only be ones that
+	// are specified by this list.
+	// If empty, it defaults to DefaultThreatLists.
+	ThreatLists []ThreatType
+
+	// RequestTimeout determines the timeout value for the http client.
+	RequestTimeout time.Duration
+
+	// Logger is an io.Writer that allows WebriskClient to write debug information
+	// intended for human consumption.
+	// If empty, no logs will be written.
+	Logger io.Writer
+
+	// compressionTypes indicates how the threat entry sets can be compressed.
+	compressionTypes []pb.CompressionType
+
+	api api
+	now func() time.Time
+}
+
+// setDefaults configures Config to have default parameters.
+// It reports whether the current configuration is valid.
+func (c *Config) setDefaults() bool {
+	if c.ServerURL == "" {
+		c.ServerURL = DefaultServerURL
+	}
+	if len(c.ThreatLists) == 0 {
+		c.ThreatLists = DefaultThreatLists
+	}
+	if c.UpdatePeriod <= 0 {
+		c.UpdatePeriod = DefaultUpdatePeriod
+	}
+	if c.RequestTimeout <= 0 {
+		c.RequestTimeout = DefaultRequestTimeout
+	}
+	if c.compressionTypes == nil {
+		c.compressionTypes = []pb.CompressionType{pb.CompressionType_RAW, pb.CompressionType_RICE}
+	}
+	return true
+}
+
+func (c Config) copy() Config {
+	c2 := c
+	c2.ThreatLists = append([]ThreatType(nil), c.ThreatLists...)
+	c2.compressionTypes = append([]pb.CompressionType(nil), c.compressionTypes...)
+	return c2
+}
+
+// WebriskClient is a client implementation of API v4.
+//
+// It provides a set of lookup methods that allows the user to query whether
+// certain entries are considered a threat. The implementation manages all of
+// local database and caching that would normally be needed to interact
+// with the API server.
+type WebriskClient struct {
+	stats  Stats // Must be first for 64-bit alignment on non 64-bit systems.
+	config Config
+	api    api
+	db     database
+	c      cache
+
+	lists map[ThreatType]bool
+
+	log *log.Logger
+
+	closed uint32
+	done   chan bool // Signals that the updater routine should stop
+}
+
+// Stats records statistics regarding WebriskClient's operation.
+type Stats struct {
```

Comment:
Seems like this should use something like OpenCensus status, rather than something bespoke.

Reply:
The stats we collect are rather simple and shown on a /status webpage of wrserver. I see why it may be preferable to use OpenCensus instead of this but I think for now our priority would be to ship this to our customers and later on iterate and make the code look pretier (we also intend to use the public libraries for webrisk instead of doing plain http requests as they are done on api.go) So what about opening an issue of modifying this to use OpenCensus and assigning that to me?